test/dag-node-test.js

```diff
@@ -32,459 +29,261 @@ module.exports = (repo) => {
     })
 
-    it('get node CID', (done) => {
-      DAGNode.create(Buffer.from('some data'), (err, node) => {
-        expect(err).to.not.exist()
-        util.cid(node, (err, cid) => {
-          expect(err).to.not.exist()
-          expect(cid.multihash).to.exist()
-          expect(cid.codec).to.equal('dag-pb')
-          expect(cid.version).to.equal(0)
-          const mh = multihash.decode(cid.multihash)
-          expect(mh.name).to.equal('sha2-256')
-          done()
-        })
-      })
+    it('get node CID with version', async () => {
+      const node = await DAGNode.create(Buffer.from('some data'))
+      const serialized = dagPB.util.serialize(node)
+      const cid = await dagPB.util.cid(serialized, { cidVersion: 0 })
+      expect(cid.multihash).to.exist()
+      expect(cid.codec).to.equal('dag-pb')
+      expect(cid.version).to.equal(0)
+      const mh = multihash.decode(cid.multihash)
+      expect(mh.name).to.equal('sha2-256')
     })
 
-    it('get node CID with hashAlg', (done) => {
-      DAGNode.create(Buffer.from('some data'), (err, node) => {
-        expect(err).to.not.exist()
-        util.cid(node, { hashAlg: 'sha2-512' }, (err, cid) => {
-          expect(err).to.not.exist()
-          expect(cid.multihash).to.exist()
-          expect(cid.codec).to.equal('dag-pb')
-          expect(cid.version).to.equal(1)
-          const mh = multihash.decode(cid.multihash)
-          expect(mh.name).to.equal('sha2-512')
-          done()
-        })
-      })
+    it('get node CID with hashAlg', async () => {
+      const node = await DAGNode.create(Buffer.from('some data'))
+      const serialized = dagPB.util.serialize(node)
+      const cid = await dagPB.util.cid(serialized, { hashAlg: multicodec.SHA2_512 })
+      expect(cid.multihash).to.exist()
+      expect(cid.codec).to.equal('dag-pb')
+      expect(cid.version).to.equal(1)
+      const mh = multihash.decode(cid.multihash)
+      expect(mh.name).to.equal('sha2-512')
     })
 
-    it('marshal a node and store it with block-service', (done) => {
-      DAGNode.create(Buffer.from('some data'), (err, node) => {
-        expect(err).to.not.exist()
-        let block
-
-        waterfall([
-          (cb) => dagPB.util.serialize(node, cb),
-          (s, cb) => dagPB.util.cid(s, (err, cid) => {
-            cb(err, {
-              buffer: s,
-              cid: cid
-            })
-          }),
-          ({ buffer, cid }, cb) => {
-            block = new Block(buffer, cid)
-            bs.put(block, cb)
-          },
-          (cb) => bs.get(block.cid, cb),
-          (retrievedBlock, cb) => {
+    it('marshal a node and store it with block-service', async () => {
+      const node = await DAGNode.create(Buffer.from('some data'))
+      const serialized = dagPB.util.serialize(node)
+      const cid = await dagPB.util.cid(serialized)
+      const block = new Block(Buffer.from(serialized), cid)
+      return new Promise((resolve) => {
```

Comment:
A good rule of thumb is that if you're ever using `new Promise` then you're probably doing something wrong.

Dealing with callback interfaces can make it tricky, but trying to manually wrap them in Promises can lead easily to errors getting lost. But that's where [`util.promisify()`](https://nodejs.org/api/util.html#util_util_promisify_original) comes in handy and it'll translate into the browser too.

Take it or leave it, but here's a patch that addresses this and makes the `BlockService` interface `await`-ready and error safe:

```patch
diff --git a/test/dag-node-test.js b/test/dag-node-test.js
index 5910e28..8fe66fd 100644
--- a/test/dag-node-test.js
+++ b/test/dag-node-test.js
@@ -1,6 +1,7 @@
 /* eslint-env mocha */
 'use strict'
 
+const { promisify } = require('util')
 const chai = require('chai')
 const chaiAsProised = require('chai-as-promised')
 const dirtyChai = require('dirty-chai')
@@ -26,7 +27,11 @@ const testBlockNamedLinks = loadFixture('test/fixtures/test-block-named-links')
 const testBlockUnnamedLinks = loadFixture('test/fixtures/test-block-unnamed-links')
 
 module.exports = (repo) => {
-  const bs = new BlockService(repo)
+  const _bs = new BlockService(repo)
+  const bs = {
+    get: promisify(_bs.get.bind(_bs)),
+    put: promisify(_bs.put.bind(_bs))
+  }
 
   describe('DAGNode', () => {
     it('create a node', async () => {
@@ -252,16 +257,9 @@ module.exports = (repo) => {
       const serialized = dagPB.util.serialize(node)
       const cid = await dagPB.util.cid(serialized)
       const block = new Block(Buffer.from(serialized), cid)
-      return new Promise((resolve) => {
-        bs.put(block, (err) => {
-          expect(err).to.not.exist()
-          bs.get(block.cid, (err, retrievedBlock) => {
-            expect(err).to.not.exist()
-            expect(retrievedBlock).to.eql(block)
-            resolve()
-          })
-        })
-      })
+      await bs.put(block)
+      const retrievedBlock = await bs.get(block.cid)
+      expect(retrievedBlock).to.eql(block)
     })
 
     it('deserialize go-ipfs block from ipldResolver', async () => {
```

Reply:
Wow thanks a lot for the code. I've used the `promisify-es6` module which is also used in js-ipfs/js-ipld which leads to a smaller browser bundle compared to the Node.js polyfill.